src/types/bytes.rs

```diff
@@ -26,6 +26,60 @@ impl PyBytes {
         unsafe { py.from_owned_ptr(ffi::PyBytes_FromStringAndSize(ptr, len)) }
     }
 
+    /// Creates a new Python bytestring object and a mutable slice to its underlying buffer.
+    /// The bytestring is uninitialised and must not be read.
+    ///
+    /// Panics if out of memory.
+    unsafe fn new_with_uninit_impl(py: Python<'_>, len: usize) -> (&mut [u8], &PyBytes) {
+        let length = len as ffi::Py_ssize_t;
+        let pyptr = ffi::PyBytes_FromStringAndSize(std::ptr::null(), length);
+        // Iff pyptr is null, py.from_owned_ptr(pyptr) will panic
+        let pybytes = py.from_owned_ptr(pyptr);
+        let buffer = ffi::PyBytes_AsString(pyptr) as *mut u8;
+        debug_assert!(!buffer.is_null());
+        let slice = std::slice::from_raw_parts_mut(buffer, len);
+        (slice, pybytes)
+    }
+
+    /// Creates a new Python bytestring object.
+    /// The `init` closure can initialise the bytestring.
+    ///
+    /// # Safety
+    /// * The bytestring is zero-initialised and can be read inside `init`.
+    ///
+    /// Panics if out of memory.
+    pub fn new_with<F: Fn(&mut [u8])>(py: Python<'_>, len: usize, init: F) -> &PyBytes {
+        let (slice, pybytes) = unsafe { Self::new_with_uninit_impl(py, len) };
+        #[cfg(feature = "slice_fill")]
+        {
+            slice.fill(0);
+        }
+        #[cfg(not(feature = "slice_fill"))]
+        {
+            slice.iter_mut().for_each(|x| *x = 0);
+        }
+        init(slice);
+        pybytes
+    }
+
+    /// Creates a new Python bytestring object.
+    ///
+    /// # Safety
+    /// * The bytestring is uninitialised and must not be read inside `init`.
+    /// * The entire bytestring must be written to inside `init` such that any code that
+    /// follows does not read uninitialised memory.
+    ///
+    /// Panics if out of memory.
+    pub unsafe fn new_with_uninit<F: Fn(&mut [u8])>(
```

Comment:
Do we need this?

Reply:
If so, the closure should take `&mut [MaybeUninit<u8>]` as @programmerjake says.

Reply:
Changed in 533c105b73d214afdb14033f45b1659cf0723233